# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели *кредитного скоринга* — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Описание данных**

- `children` — количество детей в семье;
- `days_employed` — общий трудовой стаж в днях;
- `dob_years` — возраст клиента в годах;
- `education` — уровень образования клиента;
- `education_id` — идентификатор уровня образования;
- `family_status` — семейное положение;
- `family_status_id` — идентификатор семейного положения;
- `gender` — пол клиента;
- `income_type` — тип занятости;
- `debt` — имел ли задолженность по возврату кредитов;
- `total_income` — ежемесячный доход;
- `purpose` — цель получения кредита.

## Обзор данных

Шаг 1. Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [5]:
import pandas as pd

data = pd.read_csv('../datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [7]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

Шаг 2. Выполните предобработку данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [13]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [15]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [18]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [20]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [23]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [25]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [27]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [30]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [32]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [35]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [38]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [40]:
data.duplicated().sum()

71

In [41]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [44]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [45]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [47]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [49]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [50]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Анализ данных

Шаг 3. Исследуйте данные и ответьте на вопросы

Проанализируем, влияет ли количество детей клиента, а также его семейное положение, уровень дохода и цели кредита на факт погашения кредита в срок.

### Клиенты с различным количеством детей

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Для поиска зависимости:
- Сгруппируем клиентов по количеству детей (столбец `children`).
- Сравним процент погашенных в срок кредитов в каждой группе.

In [56]:
# сгруппируем клиентов по количеству детей
# и применим агрегацию для столбца debt с целью подсчета размера групп и доли должников
data_grouped = data.groupby('children').agg({'debt': ['count', 'mean']})

# в каждой группе вычислим долю клиентов (в процентах), которые погасили кредит в срок
data_grouped['no_debt_percent'] = round((1 - data_grouped['debt']['mean']) * 100, 2)

data_grouped

debt           no_debt_percent
          count      mean                
children                                 
0         14091  0.075438           92.46
1          4808  0.092346           90.77
2          2052  0.094542           90.55
3           330  0.081818           91.82
4            41  0.097561           90.24
5             9  0.000000          100.00

<a id="result1"></a>
**Вывод:** 

Данные показывают разницу в возврате кредитов в срок клиентами с различным количеством детей:
- Лучше всех возращают кредиты клиенты с 5 детьми, а хуже всех — с 4 детьми. Однако размеры выборок клиентов с 4 детьми и 5 детьми слишком малы, чтобы опираться на этот результат.
- Клиенты без детей являются более надежными заемщиками, чем клиенты с детьми.

### Клиенты с различным семейным положением

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Для поиска зависимости:
- Сгруппируем клиентов по семейному положению (столбец `family_status`).
- Сравним процент погашенных в срок кредитов в каждой группе.

In [61]:
# сгруппируем клиентов по семейному положению 
# и применим агрегацию для столбца debt с целью подсчета размера групп и доли должников
data_grouped = data.groupby('family_status').agg({'debt': ['count', 'mean']})

# в каждой группе вычислим долю клиентов (в процентах), которые погасили кредит в срок
data_grouped['no_debt_percent'] = round((1 - data_grouped['debt']['mean']) * 100, 2)

data_grouped

debt           no_debt_percent
                       count      mean                
family_status                                         
Не женат / не замужем   2796  0.097639           90.24
в разводе               1189  0.070648           92.94
вдовец / вдова           951  0.066246           93.38
гражданский брак        4134  0.093130           90.69
женат / замужем        12261  0.075606           92.44

Из полученных результатов видно, что наиболее надежными заемщиками являются вдовцы / вдовы, а также клиенты в разводе. При этом хуже всех возращают кредиты клиенты, которые не женаты / не замужем, а также клиенты в гражданском браке.

Поскольку семейный статус клиента может быть сильно связан с возрастом, проанализируем это предположение.

In [63]:
# найдем медианный возраст клиентов, вернувших кредит в срок,
# а также медианный возраст клиентов, не вернувших кредит в срок
data[data['debt'] == 0]['dob_years'].median(), data[data['debt'] == 1]['dob_years'].median()

(43.0, 39.0)

In [64]:
# построим сводную таблицу, сгруппировав клиентов по семейному статусу
# и применив функции для подсчета доли должников и медианного возраста клиентов
data_pivot = data.pivot_table(index=['family_status'], values=['debt', 'dob_years'], 
                              aggfunc={'debt': ['count', 'mean'], 'dob_years': 'median'})

# в каждой группе вычислим долю клиентов (в процентах), которые погасили кредит в срок
data_pivot['no_debt_percent'] = round((1 - data_pivot['debt']['mean']) * 100, 2)

data_pivot

debt           dob_years no_debt_percent
                       count      mean    median                
family_status                                                   
Не женат / не замужем   2796  0.097639      36.0           90.24
в разводе               1189  0.070648      46.0           92.94
вдовец / вдова           951  0.066246      58.0           93.38
гражданский брак        4134  0.093130      41.0           90.69
женат / замужем        12261  0.075606      43.0           92.44

<a id="result2"></a>
**Вывод:** 

Данные показывают разницу в возврате кредитов в срок клиентами в группах с различным семейным положением:
- Наиболее надежными заемщиками являются вдовцы / вдовы, а также клиенты в разводе.
- Хуже всех возращают кредиты клиенты, которые не женаты / не замужем, а также клиенты в гражданском браке.
- Учитывая медианный возраст клиентов в группах, можно сделать вывод, что более надежными заемщиками являются клиенты из более "возрастных" групп (в которых клиенты успели овдоветь или развестись), менее надежными — из "молодых" (в которых клиенты еще не успели вступить в официальный брак).

### Клиенты с различным уровнем дохода

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Для поиска зависимости:
- Сгруппируем клиентов по уровню дохода (столбец `total_income_category`).
- Сравним процент погашенных в срок кредитов в каждой группе.

In [69]:
# сгруппируем клиентов по уровню дохода
# и применим агрегацию для столбца debt с целью подсчета размера групп и доли должников
data_grouped = data.groupby('total_income_category').agg({'debt': ['count', 'mean']})

# в каждой группе вычислим долю клиентов (в процентах), которые погасили кредит в срок
data_grouped['no_debt_percent'] = round((1 - data_grouped['debt']['mean']) * 100, 2)

data_grouped

debt           no_debt_percent
                       count      mean                
total_income_category                                 
A                         25  0.080000           92.00
B                       5014  0.070602           92.94
C                      15921  0.084982           91.50
D                        349  0.060172           93.98
E                         22  0.090909           90.91

<a id="result3"></a>
**Вывод:** 

Данные показывают разницу в возврате кредитов в срок клиентами с различным уровнем дохода:
- Размеры выборок клиентов с категориями доходов 'A' и 'E' (наибольший и наименьший доход соответственно) слишком малы, чтобы опираться на результаты в этих группах.
- Среди остальных групп лучше всех возвращают кредиты в срок клиенты с категорией дохода 'D' (доход от 30000 до 50000), а хуже всех — с категорией дохода 'C' (доход от 50000 до 200000)

Однако пропущенные значения ежемесячного дохода занимают существенную долю в данных (2174 записей с пропусками из 21525, то есть 10%) и угрожают достоверности исследования. Следовательно, стоит разобраться в причинах пропусков.

### Клиенты с различными целями кредита

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

Для поиска зависимости:
- Сгруппируем клиентов по цели кредита (столбец `purpose_category`).
- Сравним процент погашенных в срок кредитов в каждой группе.

In [74]:
# построим сводную таблицу, сгруппировав клиентов по цели кредита
# и применив функции для подсчета количества клиентов и доли должников в каждой группе
data_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count', 'mean'])

# в каждой группе вычислим долю клиентов (в процентах), которые погасили кредит в срок
data_pivot['no_debt_percent'] = round((1 - data_pivot['mean']['debt']) * 100, 2)

data_pivot

count      mean no_debt_percent
                           debt      debt                
purpose_category                                         
операции с автомобилем     4279  0.093480           90.65
операции с недвижимостью  10751  0.072551           92.74
получение образования      3988  0.092528           90.75
проведение свадьбы         2313  0.079118           92.09

<a id="result4"></a>
**Вывод:** 

Данные показывают разницу в возврате кредитов с различными целями:
- Лучше всех возращают кредиты, взятые на проведение операций с недвижимостью.
- Хуже всех — автокредиты и кредиты на образование (разница между ними не сильно выражена).

### Причины пропусков в данных

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

Проверим еще раз, в каких именно столбцах были пропуски, а также являются ли эти пропуски зависимыми друг от друга.

In [79]:
data_orig = pd.read_csv('../datasets/data.csv')
data_orig.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В столбцах `days_employed` и `total_income` есть по 2174 пропущенных значений. Найдем количество записей, в которых есть пропуски в двух этих столбцах одновременно.

In [81]:
# количество записей с пропусками одновременно в столбцах 'days_employed' и 'total_income'
data_orig.loc[(data_orig['days_employed'].isna()) & (data_orig['total_income'].isna())].shape[0]

2174

Значит, во всех записях с пропущенными значениями были пропущены оба столбца разом — `days_employed` и `total_income`. Такие пропуски можно отнести к неслучайным.

*Ответ:* 

В данных есть пропущенные значения в столбцах с общим трудовым стажем и ежемесячным доходом. Эти пропуски связаны между собой. В качестве возможных причин стоит рассмотреть:
- Некоторые клиенты бради кредиты без справки о доходах, следавательно, у банка просто нет информации об этих значениях.
- Технологические причины: некорректный перенос из одной базы в другую, сбой в базе данных или при выгрузке и прочее.

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ:* 

Пропуски в количественных данных заполняют характерными значениями (значениями, типичными для выборки). Для этого можно использовать средние или медианные значения. Однако **медианные значения менее зависимы от аномальных данных**, присутствующих в выборке. 

Таким образом, заполнение пропусков медианными значениями делает результаты исследования более устойчивыми к выбросам и аномалиям в данных.

## Итоги исследования

Шаг 4. Напишите общий вывод

В ходе исследования проведен анализ, влияет ли количество детей клиента, а также его семейное положение, уровень дохода и цели кредита на факт погашения кредита в срок. Анализ показал, что вышеперечисленные факторы действительно влияют на вероятность возврата кредита в срок.

**Анализ зависимости между количеством детей и возвратом кредита в срок** показал, что клиенты без детей являются более надежными заемщиками, чем клиенты с детьми.

Также был проведен **анализ зависимости между семейным положением и возвратом кредита в срок**, который показал, что наиболее надежными заемщиками являются вдовцы / вдовы, а также клиенты в разводе. При этом хуже всех возращают кредиты клиенты, которые не женаты / не замужем, а также клиенты в гражданском браке. Таким образом более надежными заемщиками являются клиенты из более "возрастных" групп, менее надежными — из "молодых".

Следом был проведен **анализ зависимости между уровнем дохода и возвратом кредита в срок**. Анализ показал, что выборки клиентов с наибольшим и наименьшим доходом (доход до 30000 и от 1000000 соответственно) слишком малы, чтобы опираться на результаты в этих группах. Среди остальных групп лучше всех возвращают кредиты в срок клиенты с доходом от 30000 до 50000, а хуже всех — с доходом от 50000 до 200000.

Кроме того, был проведен **анализ зависимости цели кредита  на его возврат в срок**. Результатом является то, что лучше всех возращают кредиты, взятые на проведение операций с недвижимостью, а хуже всех — автокредиты и кредиты на образование (разница между ними не сильно выражена).

Однако в **пропуски во входных данных могли повлиять на результаты исследования** (пропуски в столбцах `days_employed` и `total_income`). Поэтому перед использованием результатов при построении модели кредитного скоринга следует определить причины пропусков.